ydb/src/load_balancer.rs

```diff
@@ -182,6 +215,247 @@ pub(crate) async fn update_load_balancer(
+        dc_endpoints.truncate(NODES_PER_DC);
+        dc_endpoints
+    }
+
+    fn find_local_dc(&self, to_check: &[&NodeInfo]) -> YdbResult<String> {
+        let addr_to_node = self.addr_to_node(to_check);
+        if addr_to_node.is_empty() {
+            return Err(YdbError::Custom(format!("no ip addresses for endpoints")));
+        }
+        let addrs = addr_to_node.keys();
+        match self.find_fastest_address(addrs.collect()) {
+            Some(fastest_address) => Ok(addr_to_node[&fastest_address].location.clone()),
+            None => Err(YdbError::custom("could not find fastest address")),
+        }
+    }
+
+    fn addr_to_node<'a>(&'a self, nodes: &[&'a NodeInfo]) -> HashMap<String, &NodeInfo> {
+        let mut addr_to_node = HashMap::<String, &NodeInfo>::with_capacity(2 * nodes.len()); // ipv4 + ipv6
+        nodes.into_iter().for_each(|info| {
+            let host: &str;
+            let port: u16;
+            match info.uri.host() {
+                Some(uri_host) => host = uri_host,
+                None => return,
+            }
+            match info.uri.port() {
+                Some(uri_port) => port = uri_port.as_u16(),
+                None => return,
+            }
+            let _ = (host, port).to_socket_addrs().and_then(|addrs| {
+                for addr in addrs {
+                    addr_to_node.insert(addr.to_string(), info);
+                }
+                Ok(())
+            });
+        });
+        addr_to_node
+    }
+
+    fn find_fastest_address(&self, addrs: Vec<&String>) -> Option<String> {
+        let stop_measure = CancellationToken::new();
+        let (start_measure, _) = tokio::sync::broadcast::channel::<()>(1);
+        let (addr_sender, mut addr_reciever) = tokio::sync::mpsc::channel::<Option<String>>(1);
+        let addr_count = addrs.len();
+        let mut nursery = JoinSet::new();
+
+        for addr in addrs {
+            let (mut wait_for_start, stop_measure, addr, addr_sender) = (
+                start_measure.subscribe(),
+                stop_measure.clone(),
+                addr.clone(),
+                addr_sender.clone(),
+            );
+
+            nursery.spawn(async move {
+                let _ = wait_for_start.recv().await;
+                tokio::select! {
+                    connection_result = TcpStream::connect(addr.clone()) =>{
+                        match connection_result{
+                            Ok(mut connection) => {
+                                let _ = connection.shutdown().await;
+                                let _ = addr_sender.send(Some(addr)).await;
+                            },
+                            Err(_) => {let _ = addr_sender.send(None).await;},
+                        }
+                    }
+                    _ = stop_measure.cancelled() => {
+                        ();
+                    }
+                }
+            });
+        }
+
+        tokio::task::block_in_place(move || {
```

Comment:
Is it temporary block - for draft only?

Reply:
Is there any better option to keep` LoadBalancer` trait's methods synchronous? Maybe blocking channel recieve?

Reply:
Otherwise, is it ok, if I will not wait to balancer to set it's location on `set_discovery_state`, because it may cause stale information on `endpoint` call